Add 16A8W quantization configuration utility for ARM backend

Ninja91 opened the pull request:
Summary:
This diff implements a 16A8W (16-bit activations, 8-bit weights) quantization configuration utility for the ExecutorTorch ARM backend, following the feedback from D79746479.

## Key Changes

**1. New Quantization Configuration Function**
- Add `get_16a8w_quantization_config()` in `fbcode/executorch/backends/arm/quantizer/arm_quantizer.py`
- Provides 16-bit activations with HistogramObserver (better precision than 8A8W)
- Maintains 8-bit weights with MinMaxObserver/PerChannelMinMaxObserver (memory efficient)
- **Technically supported by TOSA through [EXT-INT16 extension/profile](https://www.mlplatform.org/tosa/tosa_spec.html#_conv2d)**

**2. Test Implementation**
- Add `test_linear_16a8w_tosa_INT()` test in `fbcode/executorch/backends/arm/test/ops/test_linear.py`
- Demonstrates usage of new 16A8W quantization configuration

## Benefits
- **Better Precision**: 16-bit activations provide higher precision than 8-bit. Useful for carrying precision for recurring neural nets.
- **Configurable**: Supports same parameters as existing quantization configurations

## Testing
The implementation provides the utility function and test infrastructure. Note: The test reveals that TOSA backend has limited INT16 support for some operations (view operations only support INT8/INT32/FP32/BOOL), which is expected and shows the configuration correctly produces INT16 tensors.

Differential Revision: D79763381


cc @digantdesai @freddan80 @per @zingo @oscarandersson8218

digantdesai commented on backends/arm/quantizer/arm_quantizer.py:
```diff
@@ -144,6 +144,111 @@ def get_symmetric_quantization_config(
+
+    # 16-bit activation quantization spec
+    act_quantization_spec = QuantizationSpec(
+        dtype=torch.int32,
```
Any benefit if we use int16 here? @gggekov

digantdesai replied:
@Ninja91  how does this fail if we switch to int16? Also https://github.com/pytorch/executorch/commit/8115da4afc0df46bbc5c20643926f965ba7aa74e

Ninja91 replied:
I adjusted to use int16 tosa extension and see this failure:
Seeing this error:

`
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

op_name = 'aten.view_copy.default'
tensors = [TosaArg(name='quantized_decomposed_quantize_per_tensor_default', dtype=INT16, shape=(10,), dim_order=(0,), tosa_spec=...name='aten_view_copy_default', dtype=INT16, shape=(1, 10, 1, 1), dim_order=(0, 2, 3, 1), tosa_spec=TOSA-1.0+INT+int16)]
valid_dtypes = [3, 5, 7, 1], tosa_spec = TOSA-1.0+INT+int16

    def validate_valid_dtype(
        op_name: str, tensors: Any | List[Any], valid_dtypes: Any | List[Any], tosa_spec
    ):
        """
        Validates that one or more tensors have dtypes within a set of allowed dtypes.
    
        This function checks whether the `dtype` attribute of the provided tensor(s) is one
        of the valid dtype values. It supports checking a single tensor or a list of
        tensors.
    
        Parameters:
        -----------
        op_name : str
            The name of the operation performing the validation.
        tensors : Any or List[Any]
            A tensor or list of tensors (each assumed to have `dtype` and `name` attributes)
            whose dtype will be validated.
        valid_dtypes : Any or List[Any]
            A dtype enum or list of dtype enums representing allowed dtype values.
        tosa_spec : Any
            A TosaSpecification instance indicating which TOSA version is targeted. This
            determines which serializer to use for dtype name resolution.
    
        Raises:
        -------
        ValueError
            If no tensors are provided, or if any tensor has a dtype not in `valid_dtypes`.
    
        Example:
        --------
        # Example usage:
        from executorch.backends.arm.operators.operator_validation_utils import (
            validate_valid_dtype,
        )
    
        import serializer.tosa_serializer as ts
    
        validate_valid_dtype(
            self.target,
            [*inputs, output],
            [ts.DType.INT8, ts.DType.INT32],
            output.tosa_spec,
        )
    
        """
    
        if not tensors:
            raise ValueError(
                f"{op_name}: Input tensor list is empty, cannot validate dtypes"
            )
    
        if not isinstance(valid_dtypes, List):
            valid_dtypes = [valid_dtypes]
    
        if not isinstance(tensors, List):
            tensors = [tensors]
    
        for tensor in tensors:
            if tensor.dtype not in valid_dtypes:
>               raise ValueError(
                    f"Expected tensor {tensor.name} in {op_name} to have one of the "
                    f"following dtypes: {[ts.DTypeNames[i] for i in valid_dtypes]}, "
                    f"got: {ts.DTypeNames[tensor.dtype]}"
                )
E               ValueError: Expected tensor quantized_decomposed_quantize_per_tensor_default in aten.view_copy.default to have one of the following dtypes: ['INT8', 'INT32', 'FP32', 'BOOL'], got: INT16

.../executorch/backends/arm/operators/operator_validation_utils.py:175: ValueError
`

I am trying to see if this is the only validation blocking.

per reviewed (commented):
The test cases seems to have fallen out of the last commit.
Regarding the changes to the test pipeline (in the first iteration), we also want to move to use inherit our tester from backends.test.harness instead of xnnpack tester, which also will give better possibilities to supply a custom initialized quantizer.
For the rest of the int16 support we have a few patches lined up internally to enable linear, I'll try to line them up asap.

digantdesai commented on backends/arm/quantizer/arm_quantizer.py:
```diff
@@ -143,6 +143,111 @@ def get_symmetric_quantization_config(
+        dtype=torch.int16,
+        quant_min=torch.iinfo(torch.int16).min,  # -32768
+        quant_max=torch.iinfo(torch.int16).max,  # 32767
+        qscheme=torch.per_tensor_affine,
```
so no zp?

Ninja91 replied:
I moved the tests to a different commit on top of this. Good catch on per_tensor_symmetric

Ninja91 committed:
Add 16A8W quantization configuration utility for ARM backend (#13175)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/13175

This diff implements a 16A8W (16-bit activations, 8-bit weights) quantization configuration utility for the ExecutorTorch ARM backend, following the feedback from D79746479.

## Key Changes

**1. New Quantization Configuration Function**
- Add `get_symmetric_a16w8_quantization_config()` in `fbcode/executorch/backends/arm/quantizer/arm_quantizer.py`
- Provides 16-bit activations with HistogramObserver (better precision than 8A8W)
- Maintains 8-bit weights with MinMaxObserver/PerChannelMinMaxObserver (memory efficient)
- **Technically supported by TOSA through [EXT-INT16 extension/profile](https://www.mlplatform.org/tosa/tosa_spec.html#_conv2d)**

## Benefits
- **Better Precision**: 16-bit activations provide higher precision than 8-bit. Useful for carrying precision for recurring neural nets.

Reviewed By: 3l1

Differential Revision: D79763381

Ninja91 commented:
@digantdesai , @mergennachin requested review from @robell , @gggekov . Can any of you please review this PR?
